File: python/racored/providers.py
from __future__ import annotations

import time
from dataclasses import dataclass, asdict
from typing import Any

import httpx

from .vault import CredentialVault


@dataclass(frozen=True)
class Provider:
    id: str
    name: str
    kind: str
    base_url: str
    default_model: str
    auth: str = "bearer"
    free: bool = False
    local: bool = False


PROVIDERS: dict[str, Provider] = {
    "openai": Provider("openai", "OpenAI", "responses", "https://api.openai.com/v1", "gpt-5.6-terra"),
    "anthropic": Provider("anthropic", "Anthropic", "anthropic", "https://api.anthropic.com/v1", "claude-sonnet-4-5", "x-api-key"),
    "gemini": Provider("gemini", "Google Gemini", "gemini", "https://generativelanguage.googleapis.com/v1beta", "gemini-2.5-flash", "query"),
    "openrouter": Provider("openrouter", "OpenRouter", "openai", "https://openrouter.ai/api/v1", "openrouter/auto"),
    "kimi": Provider("kimi", "Kimi / Moonshot", "openai", "https://api.moonshot.ai/v1", "kimi-k2.5"),
    "ollama": Provider("ollama", "Ollama", "ollama", "http://127.0.0.1:11434", "qwen3:8b", "none", True, True),
    "opencode": Provider("opencode", "OpenCode", "cli", "local://opencode", "configured", "local", True, True),
    "claude-code": Provider("claude-code", "Claude Code", "cli", "local://claude", "sonnet", "local", False, True),
    "kimi-code": Provider("kimi-code", "Kimi Code", "cli", "local://kimi", "default", "local", True, True),
}


class ProviderGateway:
    def __init__(self, vault: CredentialVault, timeout: float = 90):
        self.vault = vault
        self.client = httpx.AsyncClient(timeout=timeout, follow_redirects=True)

    def catalog(self) -> list[dict[str, Any]]:
        connected = set(self.vault.connected())
        return [{**asdict(provider), "connected": provider.local or provider.id in connected, "maskedKey": self.vault.masked(provider.id)} for provider in PROVIDERS.values()]

    async def health(self, provider_id: str) -> dict[str, Any]:
        provider = PROVIDERS[provider_id]
        started = time.perf_counter()
        try:
            if provider.kind == "ollama":
                response = await self.client.get(f"{provider.base_url}/api/tags", timeout=3)
                response.raise_for_status()
                models = [model.get("name") for model in response.json().get("models", [])]
                return {"ok": True, "latencyMs": round((time.perf_counter() - started) * 1000), "models": models}
            if provider.kind == "cli":
                import shutil

                executable = provider.base_url.removeprefix("local://")
                return {"ok": bool(shutil.which(executable)), "latencyMs": 0, "executable": executable}
            return {"ok": bool(self.vault.get(provider_id)), "latencyMs": 0, "configured": bool(self.vault.get(provider_id))}
        except Exception as exc:
            return {"ok": False, "latencyMs": round((time.perf_counter() - started) * 1000), "error": str(exc)}

    async def chat(self, provider_id: str, messages: list[dict[str, str]], model: str | None = None, system: str | None = None) -> dict[str, Any]:
        if provider_id not in PROVIDERS:
            raise ValueError(f"Unknown provider: {provider_id}")
        provider = PROVIDERS[provider_id]
        selected_model = model or provider.default_model
        started = time.perf_counter()
        if provider.kind == "cli":
            result = await self._cli(provider, messages)
        elif provider.kind == "ollama":
            result = await self._ollama(provider, selected_model, messages, system)
        elif provider.kind == "responses":
            result = await self._openai_responses(provider, selected_model, messages, system)
        elif provider.kind == "anthropic":
            result = await self._anthropic(provider, selected_model, messages, system)
        elif provider.kind == "gemini":
            result = await self._gemini(provider, selected_model, messages, system)
        else:
            result = await self._openai_compatible(provider, selected_model, messages, system)
        result.update({"provider": provider.id, "model": selected_model, "latencyMs": round((time.perf_counter() - started) * 1000)})
        return result

    def _key(self, provider: Provider) -> str:
        key = self.vault.get(provider.id)
        if not key:
            raise PermissionError(f"Connect {provider.name} in Racore before using it")
        return key

    async def _openai_responses(self, provider: Provider, model: str, messages: list[dict[str, str]], system: str | None) -> dict[str, Any]:
        payload: dict[str, Any] = {"model": model, "input": messages}
        if system:
            payload["instructions"] = system
        response = await self.client.post(f"{provider.base_url}/responses", headers={"Authorization": f"Bearer {self._key(provider)}"}, json=payload)
        response.raise_for_status()
        data = response.json()
        text = data.get("output_text") or "".join(part.get("text", "") for item in data.get("output", []) for part in item.get("content", []) if part.get("type") in {"output_text", "text"})
        return {"text": text, "usage": data.get("usage", {}), "rawId": data.get("id")}

    async def _openai_compatible(self, provider: Provider, model: str, messages: list[dict[str, str]], system: str | None) -> dict[str, Any]:
        payload_messages = ([{"role": "system", "content": system}] if system else []) + messages
        headers = {"Authorization": f"Bearer {self._key(provider)}", "Content-Type": "application/json", "X-Title": "Racore Browser", "HTTP-Referer": "https://racore.xyz"}
        response = await self.client.post(f"{provider.base_url}/chat/completions", headers=headers, json={"model": model, "messages": payload_messages})
        response.raise_for_status()
        data = response.json()
        return {"text": data["choices"][0]["message"].get("content", ""), "usage": data.get("usage", {}), "rawId": data.get("id")}

    async def _anthropic(self, provider: Provider, model: str, messages: list[dict[str, str]], system: str | None) -> dict[str, Any]:
        payload: dict[str, Any] = {"model": model, "max_tokens": 4096, "messages": messages}
        if system:
            payload["system"] = system
        response = await self.client.post(f"{provider.base_url}/messages", headers={"x-api-key": self._key(provider), "anthropic-version": "2023-06-01"}, json=payload)
        response.raise_for_status()
        data = response.json()
        return {"text": "".join(item.get("text", "") for item in data.get("content", []) if item.get("type") == "text"), "usage": data.get("usage", {}), "rawId": data.get("id")}

    async def _gemini(self, provider: Provider, model: str, messages: list[dict[str, str]], system: str | None) -> dict[str, Any]:
        contents = [{"role": "model" if m["role"] == "assistant" else "user", "parts": [{"text": m["content"]}]} for m in messages]
        payload: dict[str, Any] = {"contents": contents}
        if system:
            payload["systemInstruction"] = {"parts": [{"text": system}]}
        response = await self.client.post(f"{provider.base_url}/models/{model}:generateContent", params={"key": self._key(provider)}, json=payload)
        response.raise_for_status()
        data = response.json()
        text = "".join(part.get("text", "") for candidate in data.get("candidates", []) for part in candidate.get("content", {}).get("parts", []))
        return {"text": text, "usage": data.get("usageMetadata", {})}

    async def _ollama(self, provider: Provider, model: str, messages: list[dict[str, str]], system: str | None) -> dict[str, Any]:
        payload_messages = ([{"role": "system", "content": system}] if system else []) + messages
        response = await self.client.post(f"{provider.base_url}/api/chat", json={"model": model, "messages": payload_messages, "stream": False})
        response.raise_for_status()
        data = response.json()
        return {"text": data.get("message", {}).get("content", ""), "usage": {"prompt_tokens": data.get("prompt_eval_count"), "completion_tokens": data.get("eval_count")}}

    async def _cli(self, provider: Provider, messages: list[dict[str, str]]) -> dict[str, Any]:
        import asyncio

        prompt = messages[-1]["content"]
        executable = provider.base_url.removeprefix("local://")
        args = {"opencode": ["run", prompt], "claude": ["-p", prompt, "--output-format", "text"], "kimi": ["--print", prompt]}.get(executable, [prompt])
        process = await asyncio.create_subprocess_exec(executable, *args, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
        stdout, stderr = await asyncio.wait_for(process.communicate(), timeout=120)
        if process.returncode:
            raise RuntimeError(stderr.decode(errors="replace").strip() or f"{executable} failed")
        return {"text": stdout.decode(errors="replace").strip(), "usage": {}}

File: god/internal/providers/providers.go
package providers

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os/exec"
	"strings"
	"time"

	"github.com/racore/god/pkg/api"
)

type Provider struct {
	ID          string
	Name        string
	Kind        string
	BaseURL     string
	DefaultModel string
	Auth        string
	Free        bool
	Local       bool
}

var Catalog = []Provider{
	{ID: "openai", Name: "OpenAI", Kind: "responses", BaseURL: "https://api.openai.com/v1", DefaultModel: "gpt-5.6-terra", Auth: "bearer"},
	{ID: "anthropic", Name: "Anthropic", Kind: "anthropic", BaseURL: "https://api.anthropic.com/v1", DefaultModel: "claude-sonnet-4-5", Auth: "x-api-key"},
	{ID: "gemini", Name: "Google Gemini", Kind: "gemini", BaseURL: "https://generativelanguage.googleapis.com/v1beta", DefaultModel: "gemini-2.5-flash", Auth: "query"},
	{ID: "openrouter", Name: "OpenRouter", Kind: "openai", BaseURL: "https://openrouter.ai/api/v1", DefaultModel: "openrouter/auto", Auth: "bearer"},
	{ID: "kimi", Name: "Kimi / Moonshot", Kind: "openai", BaseURL: "https://api.moonshot.ai/v1", DefaultModel: "kimi-k2.5", Auth: "bearer"},
	{ID: "ollama", Name: "Ollama", Kind: "ollama", BaseURL: "http://127.0.0.1:11434", DefaultModel: "qwen3:8b", Auth: "none", Free: true, Local: true},
	{ID: "opencode", Name: "OpenCode", Kind: "cli", DefaultModel: "configured", Auth: "local", Free: true, Local: true},
	{ID: "claude-code", Name: "Claude Code", Kind: "cli", DefaultModel: "sonnet", Auth: "local", Local: true},
	{ID: "kimi-code", Name: "Kimi Code", Kind: "cli", DefaultModel: "default", Auth: "local", Free: true, Local: true},
}

type KeyFunc func(providerID string) (string, error)

type Gateway struct {
	httpClient *http.Client
	getKey     KeyFunc
}

func NewGateway(getKey KeyFunc) *Gateway {
	return &Gateway{
		httpClient: &http.Client{Timeout: 120 * time.Second},
		getKey:     getKey,
	}
}

func (g *Gateway) Catalog() []api.Provider {
	out := make([]api.Provider, len(Catalog))
	for i, p := range Catalog {
		key, _ := g.getKey(p.ID)
		masked := ""
		if key != "" {
			if len(key) < 8 {
				masked = "***"
			} else {
				masked = key[:3] + "..." + key[len(key)-4:]
			}
		}
		out[i] = api.Provider{
			ID:           p.ID,
			Name:         p.Name,
			Kind:         p.Kind,
			BaseURL:      p.BaseURL,
			DefaultModel: p.DefaultModel,
			Auth:         p.Auth,
			Free:         p.Free,
			Local:        p.Local,
			Connected:    key != "" || p.Auth == "none" || p.Auth == "local",
			MaskedKey:    masked,
		}
	}
	return out
}

func (g *Gateway) health(ctx context.Context, p Provider) (map[string]any, error) {
	switch p.Kind {
	case "ollama":
		return g.ollamaHealth(ctx)
	case "cli":
		return g.cliHealth(p)
	default:
		key, err := g.getKey(p.ID)
		if err != nil {
			return map[string]any{"ok": false, "error": err.Error()}, nil
		}
		if key == "" {
			return map[string]any{"ok": false, "error": "not connected"}, nil
		}
		return map[string]any{"ok": true}, nil
	}
}

func (g *Gateway) Health(ctx context.Context, id string) (map[string]any, error) {
	for _, p := range Catalog {
		if p.ID == id {
			return g.health(ctx, p)
		}
	}
	return nil, fmt.Errorf("provider not found: %s", id)
}

func (g *Gateway) ollamaHealth(ctx context.Context) (map[string]any, error) {
	req, _ := http.NewRequestWithContext(ctx, "GET", "http://127.0.0.1:11434/api/tags", nil)
	resp, err := g.httpClient.Do(req)
	if err != nil {
		return map[string]any{"ok": false, "error": err.Error()}, nil
	}
	defer resp.Body.Close()
	return map[string]any{"ok": resp.StatusCode == 200}, nil
}

func (g *Gateway) cliHealth(p Provider) (map[string]any, error) {
	cmd := p.ID
	if p.ID == "claude-code" {
		cmd = "claude"
	}
	if _, err := exec.LookPath(cmd); err != nil {
		return map[string]any{"ok": false, "error": fmt.Sprintf("%s not found", cmd)}, nil
	}
	return map[string]any{"ok": true}, nil
}

func (g *Gateway) Chat(ctx context.Context, providerID, model string, messages []map[string]string, system string) (*api.ChatResponse, error) {
	var p Provider
	found := false
	for _, pr := range Catalog {
		if pr.ID == providerID {
			p = pr
			found = true
			break
		}
	}
	if !found {
		return nil, fmt.Errorf("unknown provider: %s", providerID)
	}

	if model == "" {
		model = p.DefaultModel
	}

	switch p.Kind {
	case "responses":
		return g.openaiResponses(ctx, p, model, messages, system)
	case "openai":
		return g.openaiCompat(ctx, p, model, messages, system)
	case "anthropic":
		return g.anthropicChat(ctx, p, model, messages, system)
	case "gemini":
		return g.geminiChat(ctx, p, model, messages, system)
	case "ollama":
		return g.ollamaChat(ctx, p, model, messages, system)
	case "cli":
		return g.cliChat(ctx, p, model, messages, system)
	default:
		return nil, fmt.Errorf("unsupported provider kind: %s", p.Kind)
	}
}

func (g *Gateway) openaiResponses(ctx context.Context, p Provider, model string, messages []map[string]string, system string) (*api.ChatResponse, error) {
	key, err := g.getKey(p.ID)
	if err != nil {
		return nil, err
	}

	body := map[string]any{
		"model": model,
		"input": buildMessages(messages, system),
	}
	return g.doOpenAI(ctx, p.BaseURL+"/responses", key, body, p.ID, model)
}

func (g *Gateway) openaiCompat(ctx context.Context, p Provider, model string, messages []map[string]string, system string) (*api.ChatResponse, error) {
	key, err := g.getKey(p.ID)
	if err != nil {
		return nil, err
	}

	body := map[string]any{
		"model":    model,
		"messages": buildMessages(messages, system),
		"stream":   false,
	}
	return g.doOpenAI(ctx, p.BaseURL+"/chat/completions", key, body, p.ID, model)
}

func (g *Gateway) doOpenAI(ctx context.Context, url, key string, body any, providerID, model string) (*api.ChatResponse, error) {
	data, _ := json.Marshal(body)
	req, _ := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(data))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+key)

	resp, err := g.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	raw, _ := io.ReadAll(resp.Body)
	var result map[string]any
	json.Unmarshal(raw, &result)

	text := extractText(result)
	usage, _ := result["usage"]
	rawID, _ := result["id"].(string)

	return &api.ChatResponse{
		Text: text, Usage: usage, Provider: providerID,
		Model: model, RawID: rawID,
	}, nil
}

func (g *Gateway) anthropicChat(ctx context.Context, p Provider, model string, messages []map[string]string, system string) (*api.ChatResponse, error) {
	key, err := g.getKey(p.ID)
	if err != nil {
		return nil, err
	}

	anthropicMsgs := make([]map[string]any, 0)
	sysMsg := ""
	for _, m := range messages {
		role := m["role"]
		if role == "system" {
			sysMsg = m["content"]
			continue
		}
		anthropicMsgs = append(anthropicMsgs, map[string]any{
			"role":    role,
			"content": m["content"],
		})
	}
	if system != "" {
		sysMsg = system
	}

	body := map[string]any{
		"model":      model,
		"messages":   anthropicMsgs,
		"max_tokens": 4096,
	}
	if sysMsg != "" {
		body["system"] = sysMsg
	}

	data, _ := json.Marshal(body)
	req, _ := http.NewRequestWithContext(ctx, "POST", p.BaseURL+"/messages", bytes.NewReader(data))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-api-key", key)
	req.Header.Set("anthropic-version", "2023-06-01")

	resp, err := g.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	raw, _ := io.ReadAll(resp.Body)
	var result map[string]any
	json.Unmarshal(raw, &result)

	text := ""
	if content, ok := result["content"].([]any); ok && len(content) > 0 {
		if first, ok := content[0].(map[string]any); ok {
			text, _ = first["text"].(string)
		}
	}
	usage, _ := result["usage"]

	return &api.ChatResponse{Text: text, Usage: usage, Provider: p.ID, Model: model}, nil
}

func (g *Gateway) geminiChat(ctx context.Context, p Provider, model string, messages []map[string]string, system string) (*api.ChatResponse, error) {
	key, err := g.getKey(p.ID)
	if err != nil {
		return nil, err
	}

	geminiContents := make([]map[string]any, 0)
	for _, m := range messages {
		role := "user"
		if m["role"] == "assistant" || m["role"] == "model" {
			role = "model"
		}
		geminiContents = append(geminiContents, map[string]any{
			"role": role,
			"parts": []map[string]any{
				{"text": m["content"]},
			},
		})
	}

	body := map[string]any{"contents": geminiContents}
	url := fmt.Sprintf("%s/models/%s:generateContent", p.BaseURL, model)

	data, _ := json.Marshal(body)
	req, _ := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(data))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-goog-api-key", key)

	resp, err := g.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	raw, _ := io.ReadAll(resp.Body)
	var result map[string]any
	json.Unmarshal(raw, &result)

	text := ""
	if candidates, ok := result["candidates"].([]any); ok && len(candidates) > 0 {
		if first, ok := candidates[0].(map[string]any); ok {
			if content, ok := first["content"].(map[string]any); ok {
				if parts, ok := content["parts"].([]any); ok && len(parts) > 0 {
					if part, ok := parts[0].(map[string]any); ok {
						text, _ = part["text"].(string)
					}
				}
			}
		}
	}

	return &api.ChatResponse{Text: text, Provider: p.ID, Model: model}, nil
}

func (g *Gateway) ollamaChat(ctx context.Context, p Provider, model string, messages []map[string]string, system string) (*api.ChatResponse, error) {
	ollamaMsgs := buildMessages(messages, system)
	body := map[string]any{
		"model":    model,
		"messages": ollamaMsgs,
		"stream":   false,
	}

	data, _ := json.Marshal(body)
	req, _ := http.NewRequestWithContext(ctx, "POST", p.BaseURL+"/api/chat", bytes.NewReader(data))
	req.Header.Set("Content-Type", "application/json")

	resp, err := g.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	raw, _ := io.ReadAll(resp.Body)
	var result map[string]any
	json.Unmarshal(raw, &result)

	text := ""
	if msg, ok := result["message"].(map[string]any); ok {
		text, _ = msg["content"].(string)
	}

	return &api.ChatResponse{Text: text, Provider: p.ID, Model: model}, nil
}

func (g *Gateway) cliChat(ctx context.Context, p Provider, model string, messages []map[string]string, system string) (*api.ChatResponse, error) {
	prompt := system
	if prompt != "" {
		prompt += "\n\n"
	}
	for _, m := range messages {
		prompt += m["role"] + ": " + m["content"] + "\n"
	}

	var c *exec.Cmd
	switch p.ID {
	case "opencode":
		c = exec.CommandContext(ctx, "opencode", "run", prompt)
	case "claude-code":
		c = exec.CommandContext(ctx, "claude", "-p", prompt)
	case "kimi-code":
		c = exec.CommandContext(ctx, "kimi", "--print", prompt)
	default:
		return nil, fmt.Errorf("unknown CLI provider: %s", p.ID)
	}

	out, err := c.Output()
	if err != nil {
		return nil, fmt.Errorf("CLI error: %w", err)
	}

	return &api.ChatResponse{
		Text: strings.TrimSpace(string(out)), Provider: p.ID, Model: model,
	}, nil
}

func buildMessages(messages []map[string]string, system string) []map[string]any {
	out := make([]map[string]any, 0)
	if system != "" {
		out = append(out, map[string]any{"role": "system", "content": system})
	}
	for _, m := range messages {
		out = append(out, map[string]any{"role": m["role"], "content": m["content"]})
	}
	return out
}

func extractText(result map[string]any) string {
	if output, ok := result["output"].(map[string]any); ok {
		if text, ok := output["text"].(string); ok {
			return text
		}
	}
	if choices, ok := result["choices"].([]any); ok && len(choices) > 0 {
		if first, ok := choices[0].(map[string]any); ok {
			if msg, ok := first["message"].(map[string]any); ok {
				if text, ok := msg["content"].(string); ok {
					return text
				}
			}
			if text, ok := first["text"].(string); ok {
				return text
			}
		}
	}
	return ""
}
